src/sql/base/browser/ui/table/treeGrid.ts

```diff
@@ -0,0 +1,167 @@
+function defaultTreeGridFilter<T extends Slick.SlickData>(data: T[], columns: FilterableColumn<T>[], cellValueGetter: CellValueGetter = defaultCellValueGetter): T[] {
+	let filteredData = defaultFilter(data, columns, cellValueGetter);
+
+	// filtering out rows which have parent/grandparents collapsed.
+	filteredData = filteredData.filter((item) => {
+		let parent = data[item.parent];
+		while (parent) {
+			if (!parent.expanded) {
+				return false;
+			}
+			parent = data[parent.parent];
+		}
+		return true;
+	});
+
+	return filteredData;
+}
+
+/**
+ * TreeGrid component displays a hierarchical table data grouped into expandable and collapsible nodes.
+ */
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+		this._tableContainer.setAttribute('role', 'treegrid');
+
+		if (configuration?.dataProvider && configuration.dataProvider instanceof TableDataView) {
+			this._data = configuration.dataProvider;
+		} else {
+			this._data = new TableDataView<T>(configuration && configuration.dataProvider as Array<T>,
+				undefined,
+				undefined,
+				defaultTreeGridFilter,
+				undefined);
+		}
+
+		this._grid.onClick.subscribe((e, data) => {
+			this.setRowExpandedState(data.row);
+			return false;
+		});
+
+		// The events returned by grid are Jquery events. These events can be handled by returning false which executes preventDefault and stopPropagation
+		this._grid.onKeyDown.subscribe((e, data) => {
+			const keyboardEvent = (<any>e).originalEvent;
+			if (keyboardEvent instanceof KeyboardEvent) {
+				let event = new StandardKeyboardEvent(keyboardEvent);
+				if (event.keyCode === KeyCode.Enter) {
+					// toggle the collapsed state of the row
+					this.setRowExpandedState(data.row);
+					return false;
+				} else if (event.keyCode === KeyCode.LeftArrow) {
+					// Left arrow on first cell of the expanded row collapses it
+					if (data.cell === 0) {
+						this.setRowExpandedState(data.row, false); // Collapsing state
+						return false;
+					}
+				} else if (event.keyCode === KeyCode.RightArrow) {
+					// Right arrow on last cell of the collapsed row expands it.
+					if (data.cell === (this._grid.getColumns().length - 1)) {
+						this.setRowExpandedState(data.row, true);
+						return false;
+					}
+				}
+			}
+			return true;
+		});
+
+		this._grid.onRendered.subscribe((e, data) => {
+			const visibleRows = this._grid.getViewport();
+			for (let i = visibleRows.top; i < visibleRows.bottom; i++) {
+				const rowData = this._data.getItem(i);
+				// Getting the row div that corresponds to the data row
+				const rowElement = this._tableContainer.querySelector(`div [role="row"][aria-rowindex="${(i + 1)}"]`);
+				// If the row element is found in the dom, we are setting the required aria attributes for it.
+				if (rowElement) {
+					if (rowData.expanded !== undefined) {
```

Comment:
have you confirmed that slickgrid is not reusing the row container? if it does reuse, then not all rowdata has expanded set, and there will be a problem.

Reply:
https://github.com/microsoft/SlickGrid.ADS/blob/08f1ca3f6eb2cd1ea7f56e2af4b69600d48e1932/slick.grid.js#L2390

Slickgrid used a different element for every row. It might store those elements in the cache, but it will still only reuse it for the same row.

Reply:
I will still unset the attributes if the properties are undefined. In that way we should not see any such issues.